sdk/servicebus/service-bus/src/core/linkEntity.ts

```diff
@@ -262,6 +262,12 @@ export abstract class LinkEntity<LinkT extends Receiver | AwaitableSender | Requ
       this.checkIfConnectionReady();
 
       logger.verbose(`${this._logPrefix} Creating with options %O`, options);
+      if (this.link) {
+        // For the cases such as connection is disconnected, onAmqpClose and onSessionClose events,
+        // we don't clear the link for sender. To compensate for the above,
+        // we are calling close on the link to cleanup before creating the new one.
+        await this.link.close();
```

Comment:
Why `close()` and not `closeLink()`?
The former sets the `_wasClosedPermanently ` to true and removes the link from the context cache which may have other repercussions

Reply:
`.close()` is different from `.link.close()`, we are not calling `.close()` here.

`closeLink` would try acquiring a lock but the initialization already has it, leading into a deadlock. Hence `closeLink` shouldn't be called. Alternatively, `closeLinkImpl` can be called which is a combination of clear-timeout and `link.close()`. Timers are handled in ensureTokenRenewal anyway, so `link.close()` makes sense here.

Reply:
Anyways, reverted this in favour of the proposal for cleanup with detached where onDetached would be called before the `refreshConnection()` is invoked.